src/dune_rules/scope.ml

```diff
@@ -26,158 +26,320 @@ module DB = struct
   module Found_or_redirect : sig
     type t = private
       | Found of Lib_info.external_
-      | Redirect of (Loc.t * Lib_name.t)
+      | Redirect of
+          { loc : Loc.t
+          ; to_ : Lib_name.t
+          ; enabled : Toggle.t Memo.Lazy.t
+          }
+      | Deprecated_library_name of (Loc.t * Lib_name.t)
 
-    val redirect : Lib_name.t -> Loc.t * Lib_name.t -> Lib_name.t * t
+    val redirect
+      :  enabled:Toggle.t Memo.Lazy.t
+      -> Lib_name.t
+      -> Loc.t * Lib_name.t
+      -> Lib_name.t * t
+
+    val deprecated_library_name : Lib_name.t -> Loc.t * Lib_name.t -> Lib_name.t * t
     val found : Lib_info.external_ -> t
   end = struct
     type t =
       | Found of Lib_info.external_
-      | Redirect of (Loc.t * Lib_name.t)
+      | Redirect of
+          { loc : Loc.t
+          ; to_ : Lib_name.t
+          ; enabled : Toggle.t Memo.Lazy.t
+          }
+      | Deprecated_library_name of (Loc.t * Lib_name.t)
+
+    let redirect ~enabled from (loc, to_) =
+      if Lib_name.equal from to_
+      then Code_error.raise ~loc "Invalid redirect" [ "to_", Lib_name.to_dyn to_ ]
+      else from, Redirect { loc; to_; enabled }
+    ;;
 
-    let redirect from (loc, to_) =
+    let deprecated_library_name from (loc, to_) =
       if Lib_name.equal from to_
       then Code_error.raise ~loc "Invalid redirect" [ "to_", Lib_name.to_dyn to_ ]
-      else from, Redirect (loc, to_)
+      else from, Deprecated_library_name (loc, to_)
     ;;
 
     let found x = Found x
   end
 
+  let resolve_name =
+    let module Resolve_result = Lib.DB.Resolve_result in
+    let module With_multiple_results = Resolve_result.With_multiple_results in
+    fun ~resolve_sentinel id_map name ->
+      match
+        Lib_name.Map.find id_map name |> Option.map ~f:Lib_info.Sentinel.Set.to_list
+      with
+      | None ->
+        Memo.return (With_multiple_results.resolve_result Resolve_result.not_found)
+      | Some [] -> assert false
```

Comment:
I think there's something like `Code_error` that could be used here instead of plain assert?